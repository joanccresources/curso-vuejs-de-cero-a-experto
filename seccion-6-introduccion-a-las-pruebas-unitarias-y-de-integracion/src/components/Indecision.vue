<template>
  <img v-if="image" :src="image" alt="Movies" class="img-bg" />
  <div class="bg-dark"></div>
  <div class="content">
    <div class="container">
      <div class="row">
        <div class="col-12">
          <input
            type="text"
            class="form-control"
            placeholder="Hazme una pregunta"
            v-model="question"
          />
          <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>
          <div class="mt-5" v-if="isValidQuestion">
            <p>{{ question }}</p>
            <p class="h1">{{ answer }}</p>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: "Indecision",
  props: {},
  data() {
    return {
      question: "",
      answer: null,
      image: null,
      isValidQuestion: false,
    };
  },
  methods: {
    async getAnswer() {
      try {
        this.answer = "Pensando...";
        const res = await fetch("https://yesno.wtf/api");
        const { answer, image } = await res.json();
        this.answer = answer === "yes" ? "Si" : "No!";
        this.image = image;
      } catch (error) {
        console.log("IndecisionComponent: ", error);
        this.answer = "No se pudo cargar del API";
        this.image = null;
      }
    },
  },
  // Watch - puede estar pendiente de las propiedades de data
  // Es como un observable
  // Recibe 2 valores
  watch: {
    // La funcion recibe el nombre de la propiedad
    question(value, oldValue) {
      this.isValidQuestion = false;
      console.log({ value });
      if (!value.includes("?")) return;

      this.isValidQuestion = true;
      // TODO: Realizar peticion HTTP
      this.getAnswer();
    },
  },
};
</script>

<style lang="scss" scoped>
.img-bg {
  max-width: 100%;
  width: 100%;
  height: 100vh;
  object-fit: cover;
  position: fixed;
  inset: 0;
  z-index: 2;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4) !important;
  position: fixed;
  inset: 0;
  z-index: 2;
}

.content {
  position: relative;
  z-index: 9;
  color: white;
}
</style>
