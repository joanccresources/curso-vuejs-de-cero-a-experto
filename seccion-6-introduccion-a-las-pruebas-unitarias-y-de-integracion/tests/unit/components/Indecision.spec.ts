import { VueWrapper, shallowMount } from "@vue/test-utils";
import { ComponentPublicInstance } from "vue";
import Indecision from "@/components/Indecision.vue";

describe("Indecision Component", () => {
  let wrapper: VueWrapper<any, ComponentPublicInstance<{}, any>>,
    cslSpy: jest.SpyInstance<void, any[]>;

  // solucion a error con el fetch
  // Recuerda poner en tsconfig "node"
  // "types": ["node", "webpack-env", "jest"],
  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          answer: "yes",
          forced: false,
          image: "https://yesno.wtf/assets/yes/2.gif",
        }),
    })
  ) as jest.Mock;

  beforeEach(() => {
    wrapper = shallowMount(Indecision);
    // Quiero espiar o "hacer un mok" a console.log
    cslSpy = jest.spyOn(console, "log");

    // Para limpiar los moks
    jest.clearAllMocks();
  });

  test("debe hacer match con el snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("escribir en el input no debe disparar nada (console.log)", async () => {
    // Para llegar al metodo getAnswer necesitamos "wrapper.vm"
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");

    const input = wrapper.find("input");
    // Escribimos en en el input "Hola Mundo"
    await input.setValue("Hola Mundo");

    // Queremos que el console.log halla sido llamado
    // expect(cslSpy).toHaveBeenCalled();

    // Tambien para ver que funcione deberia llamarse solo una vez
    expect(cslSpy).toHaveBeenCalledTimes(1);

    // Se debio llamar 0 veces ya que no tiene "?"
    expect(getAnswerSpy).toHaveBeenCalledTimes(0);

    // opcional a ⬆️ tambien seria valido - No fue llamado
    // expect(getAnswerSpy).not.toHaveBeenCalled();
  });

  test("escribir el simbolo '?' debe disparar el getAnswer", async () => {
    // Espiamos la el method "getAnswer"
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");

    // Le agregamos al input "?"
    const input = wrapper.find("input");
    await input.setValue("eres?");

    expect(cslSpy).toHaveBeenCalledTimes(1);
    // Deberia ejecutarse al menos 1 vez ya que tiene "?"
    expect(getAnswerSpy).toHaveBeenCalledTimes(1);
  });

  test(`pruebas en getAnswer`, async () => {
    // Esperamos por ser asincrona
    await wrapper.vm.getAnswer();

    console.log(wrapper.vm.image); // https://yesno.wtf/assets/yes/2.gif
    console.log(wrapper.vm.answer); // Si

    // Entonces compruebo que la imagen existe
    const img = wrapper.find("img");
    expect(img.exists()).toBeTruthy();
    // El valor de la imagen debe ser "https://yesno.wtf/assets/yes/2.gif"
    // porque asi lo pusimos
    expect(wrapper.vm.image).toBe("https://yesno.wtf/assets/yes/2.gif");
    // Lo mismo para el answer
    expect(wrapper.vm.answer).toBe("Si");
  });

  test(`pruebas en getAnswer - Fallo en el API`, async () => {
    (global.fetch as jest.Mock).mockImplementationOnce(() =>
      Promise.reject("Api is down")
    );
    
    // Entonces
    await wrapper.vm.getAnswer();
    const img = wrapper.find("img");
    expect(img.exists()).toBeFalsy();
    expect(wrapper.vm.answer).toBe("No se pudo cargar del API");
  });
});
