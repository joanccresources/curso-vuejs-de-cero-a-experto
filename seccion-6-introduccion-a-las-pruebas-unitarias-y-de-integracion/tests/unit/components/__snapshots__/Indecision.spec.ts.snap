// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component debe hacer match con el snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="content">
  <div class="container">
    <div class="row">
      <div class="col-12"><input type="text" class="form-control" placeholder="Hazme una pregunta">
        <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>
        <!--v-if-->
      </div>
    </div>
  </div>
</div>
`;
